### DISCUSSION TITLE
Updating Matrix math objects

### DISCUSSION THREAD

**Discussion by pstaabp** [openwebwork/pg#991] [Ideas]
I've been writing/rewriting a bunch of linear algebra problems lately and have a few things that I'd like to add to the Matrix math object. 
* a submatrix routine that either specifies the rows/cols as array refs to extract or row/col as integer to eliminate from the original matrix.
* a setElement, setRow, setColumn routine.  My original thought is to update the current matrix (not return a copy of the original). 
* a test suite for these routines.

There was discussion at some point to add other things to the Matrix, for example some of the routines from MatrixReduce.pl for row operations.  

Thoughts?  I have started this, but wondered if the row operations should be added to this or keep a slim PR.

**Comment by Alex-Jordan** [comment]
Until today, I did not know about several matrix-related macro files. My bad for not looking previously. But there are these:

* `contexts/contextMatrixExtras.pl`
* `math/MatrixCheckers.pl`
* `math/MatrixReduce.pl`
* `math/MatrixUnits.pl`
* `math/PGmatrixmacros.pl`
* `math/PGmorematrixmacros.pl`

For some of what these offer, I wonder if they should just be in `Value::Matrix` to begin with.

**Comment by Alex-Jordan** [comment]
Here is a list of what those macros offer:

* `det()`, `tr()`, and `^T` understood in Formulas.
* special checkers for sets of vectors (e.g. to see if they form a basis of whatever set is spanned by the official answers). These are based on `parserMultiAnswer.pl`, so you have to give away to the student how many vectors are in the basis, because there is a separate answer input field (or array) for each vector. Note this comes from "MatrixCheckers" but it's really more about vector collections.
* rref with floating point, rref with fractions, set value of one entry in an existing matrix [in @pstaabp's list], construct elementary matrices [in my recent PR], apply elementary row operations to a matrix
* generate 2x2, 3x3, or 4x4 (random?) matrices that have determinant 1 or -1
* stuff that is more for working with perl nested array references than MathObject matrices

In addition to what's been mentioned so far, one more thing I would like is to specify m, n for an mxn matrix, and specify p for a number of pivots. And then generate a random rref matrix under those conditions, with free entries constrained in some range. I would like that matrix math object to have a property that tells me where the pivot positions are.

In the vectors arena, I'd like a simple test for if some vector b is in the span of a given list of vectors.

Given matrices A (mxn) and B (mxk) I'd like to be able to augment A with B, and get [A|B]. And related, I'd like for a matrix `$M` to have a property or two for augmentation lines. Most commonly just one vertical augmentation line, but it wouldn't hurt to also allow for a horizontal one. And also for the possibility of more than one in each direction.

It seems that the only utility for solving Ax=b that we have requires A to be square. (Am I wrong about that?) I'd like to be able to provide an A and a b, and get back an appropriate thing. No solutions => `[]`. One solution => `[v]`. Many solutions => `[v,[b1,b2,...]]`.

I guess technically a MathObject Matrix can be 1D or 3D, 4D, etc, and I'm just thinking 2D. I don't know if that messes with many of the things I am thinking about getting.

**Comment by somiaj** [comment]
Note, if you are in the process of adding more functionality to the Value::Matrix class, augmented matrices would be useful too. I have my own context for this that you are welcome to use as a base. It works (but needs some clean up and has a few minor issues with working with normal matrices). https://github.com/somiaj/pgMacros/blob/main/contextAugmentedMatrix.pl

**Comment by drgrice1** [comment]
I also have a macro for augmented matrices.  It is similar to @somiaj's macro.  It probably has similar issues that he noted for his macro, and is not developed to the point of being ready to submit to PG, but works for my purposes.  It is attached.
[contextAMatrix.pl.txt](https://github.com/openwebwork/pg/files/14280779/contextAMatrix.pl.txt)

**Comment by drgrice1** [comment]
Note that my version is a bit more limited than @somiaj's in that it assumes there is only one column in the augmentation part of the matrix (because that is all I needed!).

**Comment by taniwallach** [comment]
Just for the record - the current basis_checker algorithms in `MatrixCheckers.pl` are *not* reliable. See https://github.com/openwebwork/pg/issues/409

I write new ones, but as way back when I did not find someone to help me testing and verifying that they are "better" they are stuck as code in my problem directories.

The attached ZIP file has the current version of the basis checkers I use, as well as macros for partial credit for matrices with some correct entries. These are probably somewhat updated compared to the files in https://github.com/openwebwork/pg/issues/409.

[BasisCheckers.zip](https://github.com/openwebwork/pg/files/14343499/BasisCheckers.zip)
